psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1504 @@
+            double e_linear_temp = 0.0;
+
+            for (int u = 0; u < lmo_pao_dr[i].size(); u++) {
+                for (int v = 0; v < lmo_pao_dr[j].size(); v++) {
+
+                    Vector3 iu = lmo_pao_dr[i][u];
+                    Vector3 jv = lmo_pao_dr[j][v];
+
+                    double num_actual = iu.dot(jv) - 3 * (iu.dot(Rh_ij) * jv.dot(Rh_ij)); 
+                    num_actual *= num_actual;
+
+                    double num_linear = -2 * iu.dot(jv);
+                    num_linear *= num_linear;
+
+                    double denom = (lmo_pao_e[i]->get(u) + lmo_pao_e[j]->get(v)) - (F_lmo->get(i,i) + F_lmo->get(j,j));
+
+                    e_actual_temp += (num_actual / denom);
+                    e_linear_temp += (num_linear / denom);
+              }
+            }
+
+            e_actual_temp *= (-4 * pow(R_ij.norm(), -6));
+            e_linear_temp *= (-4 * pow(R_ij.norm(), -6));
+
+            e_actual->set(i, j, e_actual_temp);
+            e_actual->set(j, i, e_actual_temp);
+
+            e_linear->set(i, j, e_linear_temp);
+            e_linear->set(j, i, e_linear_temp);
+
+        }
+    }
+
+}
+
+
+void DLPNOMP2::sparsity_prep() {
+
+    int natom = molecule_->natom();
+    int nbf = basisset_->nbf();
+    int nshell = basisset_->nshell();
+    int naux = ribasis_->nbf();
+    int naocc = nalpha_ - basisset_->n_frozen_core(options_.get_str("FREEZE_CORE"), molecule_);
+
+    auto bf_to_atom_ = std::vector<int>(nbf);
+    auto ribf_to_atom_ = std::vector<int>(naux);
+
+    for(size_t i = 0; i < nbf; ++i) {
+        bf_to_atom_[i] = basisset_->function_to_center(i);
+    }
+
+    for(size_t i = 0; i < naux; ++i) {
+        ribf_to_atom_[i] = ribasis_->function_to_center(i);
+    }
+
+    outfile->Printf("  ==> Forming Local MO Domains <==\n");
+
+    // map from LMO to local DF domain (aux basis functions)
+    // locality determined via mulliken charges
+
+    lmo_to_ribfs_.resize(naocc);
+    lmo_to_riatoms_.resize(naocc);
+
+    for (size_t i=0; i < naocc; ++i) {
+
+        // atomic mulliken populations for this orbital
+        std::vector<double> mkn_pop(natom, 0.0);
+
+        SharedMatrix P_i = reference_wavefunction_->S()->clone();
+
+        for(size_t u=0; u < nbf; u++){
+            P_i->scale_row(0, u, C_lmo->get(u, i));
+            P_i->scale_column(0, u, C_lmo->get(u, i));
+        }
+
+        for(size_t u=0; u < nbf; u++){
+            int centerU = basisset_->function_to_center(u);
+            double p_uu = P_i->get(u,u);
+
+            for(size_t v=0; v < nbf; v++){
+                int centerV = basisset_->function_to_center(u);
+                double p_vv = P_i->get(v,v);
+
+                // off-diag pops (p_uv) split between u and v prop to diag pops
+                double p_uv = P_i->get(u,v);
+                mkn_pop[centerU] += p_uv * ((p_uu) / (p_uu + p_vv));
+                mkn_pop[centerV] += p_uv * ((p_vv) / (p_uu + p_vv));
+
+            }
+        }
```

Comment:
I'm wondering whether anyone has done this with better methods, e.g. IAO. Mulliken doesn't even have a basis set limit...

Reply:
The other part of the algorithms should work just the same, regardless of how the partial charge is computed.

Reply:
Turbomole's OSV-PNO uses IAOs. Maybe Molpro's variant as well.

ORCA had initially some problems with finding different LMO solutions, e.g. for augmented basis sets, and worked on getting a really robust localizer (augmented hessian algorithm). 

So looking into what LMOs PSI4 should use is something worth spending time on. Long term, not this PR.

Reply:
IAOs are a simple solution; they're almost a drop-in replacement to Mulliken. One can also come up with other mathematically well-defined partial charges, as we discuss in https://pubs.acs.org/doi/10.1021/ct401016x

Reply:
That's a cool paper Susi, thanks. IAO charges would be an improvement over Mulliken charges; I know Molpro's related PNO-LMP2 method uses IAO charges for domain construction. There's some IAO/IBO code buried in the `fisapt` module. Moving that code to the globally accessible `Localizer` class would be a worthwhile project.

Thanks for the info about the augmented hessian Holger. I did notice problems with the localized orbitals when using a lot of diffuse functions, so that might be another good extension.